### PR TITLE
Add support for the rodata op in the VM to EmitC conversion

### PR DESCRIPTION
**Pull request by simon-camp**

### CODE REVIEW

**Review by marbre — APPROVED** [review]
LGTM with now reusing the `PreserveCompilerHints` parts. Let's see what @benvanik says :)

**Comment by marbre** [comment]
FYI, you need to rebase as `emitc/Dialect/EmitC/EmitCDialect.h` was moved to `emitc/Dialect/EmitC/IR/EmitC.h`. This was integrated into IREE with commit 4287062fca9f27a7178a2cbffcddf1defb20dabb.

File: iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp [hunk 1]
```diff
@@ -135,8 +177,19 @@ static LogicalResult initializeGlobals(IREE::VM::ModuleOp moduleOp,
...
     }
   }
+  // TODO(simon-camp): Support globals with different element type
 
-  // TODO(simon-camp): Support vm.global.i64 and vm.global.ref
+  for (auto rodataOp : moduleOp.getOps<IREE::VM::RodataOp>()) {
+    std::string buffer_name =
+        moduleOp.getName().str() + "_" + rodataOp.getName().str();
+    output << "iree_vm_buffer_initialize("
+           << "IREE_VM_BUFFER_ACCESS_ORIGIN_HOST, "
```
Comment (benvanik) [inline iree/compiler/Dialect/VM/Target/C/CModuleTarget.cpp:186]: this should be `IREE_VM_BUFFER_ACCESS_ORIGIN_MODULE` - that indicates that the buffer will only be available for as long as the module is loaded/initialized

File: iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        return emitError(loc) << "unhandled element bitwidth " << bitwidth;
+    }
+  } else {
+    return emitError(loc) << "unimplemented attribute encoding: "
+                          << elementsAttr.getType();
+  }
+
+  return success();
+}
+}  // namespace VM
```
Comment (benvanik) [inline iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.cpp:128]: style: newline between function and } // namespace block

File: iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace VM {
+LogicalResult serializeConstantArray(Location loc, ElementsAttr elementsAttr,
```
Comment (benvanik) [inline iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.h:18]: style: newlines between namespace stuff and contents

File: iree/vm/ops.h [hunk 4]
```diff
@@ -207,7 +216,9 @@ static inline int32_t vm_cmp_lte_f32o(float lhs, float rhs) {
 static inline int32_t vm_cmp_lte_f32u(float lhs, float rhs) {
   return (isunordered(lhs, rhs) || islessequal(lhs, rhs)) ? 1 : 0;
 }
-static inline int32_t vm_cmp_nan_f32(float operand) { return isnan(operand); }
+static inline int32_t vm_cmp_nan_f32(float operand) {
+  return isnan(operand) ? 1 : 0;
```
Comment (benvanik) [inline iree/vm/ops.h:220]: good catch!

**Review by benvanik — CHANGES_REQUESTED** [review]
awesome! this unlocks a lot! just one minor tweak and a style nit then can land

File: iree/vm/test/ref_ops.mlir [hunk 5]
```diff
@@ -0,0 +1,41 @@
+vm.module @ref_ops {
+    vm.rodata private @buffer_i8 dense<[1, 2, 3]> : tensor<3xi8>
```
Comment (benvanik) [inline iree/vm/test/ref_ops.mlir:2]: style: 2 space indent

File: iree/vm/test/emitc/module_test.cc [hunk 6]
```diff
@@ -55,7 +57,7 @@ std::vector<TestParams> GetModuleTestParams() {
       {arithmetic_ops_descriptor_, arithmetic_ops_create},
       {arithmetic_ops_f32_descriptor_, arithmetic_ops_f32_create},
       {arithmetic_ops_i64_descriptor_, arithmetic_ops_i64_create},
-      {assignment_ops_descriptor_, assignment_ops_create},
+      // {assignment_ops_descriptor_, assignment_ops_create},
```
Comment (benvanik) [inline iree/vm/test/emitc/module_test.cc:60]: intended?

Comment (benvanik) [reply iree/vm/test/emitc/module_test.cc:60]: (nevermind see the todo in the build file)

File: iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+#include "mlir/Support/LogicalResult.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace VM {
+
+LogicalResult serializeConstantArray(Location loc, ElementsAttr elementsAttr,
+                                     size_t alignment, uint8_t *dst);
+}  // namespace VM
```
Comment (benvanik) [inline iree/compiler/Dialect/VM/Target/ConstantEncodingUtils.h:21]: nit: you got the top, but not the bottom :P

**Review by benvanik — APPROVED** [review]
two small style nits and otherwise looks great!

**Pull request merged**